astropy/io/votable/dataorigin.py

```diff
@@ -0,0 +1,518 @@
+            if info_name == "standardid":
+                if not data_origin.query.service_protocol:
+                    if data_origin.info is None:
+                        data_origin.infos = []
+                    data_origin.quey.infos.append(info)
+                    data_origin.query.service_protocol = info.value
+
+
+def __extract_info_from_table(
+    table: astropy.io.votable.tree.Table, data_origin: DataOrigin
+):
+    """(internal) extract and populate dataOrigin from astropy.io.votable.tree.Table
+
+    Parameters
+    ----------
+    table : astropy.io.votable.tree.Table
+        Table to explore
+
+    data_origin : DataOrigin
+        container to fill.
+    """
+    __extract_generic_info(table, table.infos, data_origin)
+
+
+def __extract_info_from_resource(
+    resource: astropy.io.votable.tree.Resource,
+    data_origin: DataOrigin,
+    recursive: bool = True,
+):
+    """(internal) extract and populate dataOrigin from astropy.io.votable.tree.Resource
+
+    Parameters
+    ----------
+    param resource : astropy.io.votable.tree.Resource
+        Resource to explore.
+
+    data_origin : DataOrigin
+        container to fill.
+
+    recursive : bool, optional
+        make a recursive search (default: True)
+    """
+    __extract_generic_info(resource, resource.infos, data_origin)
+    __extract_dali_info(resource.infos, data_origin)
+    if recursive:
+        for table in resource.tables:
+            __extract_info_from_table(table, data_origin)
+
+
+def __extract_info_from_votable(
+    votable: astropy.io.votable.tree.VOTableFile,
+    data_origin: DataOrigin,
+    recursive: bool = True,
+):
+    """(internal) extract and populate dataOrigin from astropy.io.votable.tree.VOTableFile
+
+    Parameters
+    ----------
+    votable : astropy.io.votable.tree.VOTableFile
+        VOTableFile to explore
+
+    data_origin : DataOrigin
+        container to fill
+
+    recursive : bool, optional
+        make a recursive search (default: True)
+    """
+    __extract_generic_info(votable, votable.infos, data_origin)
+    if recursive:
+        for resource in votable.resources:
+            __extract_info_from_resource(resource, data_origin)
+
+
+def extract_data_origin(vot_element: astropy.io.votable.tree.Element) -> DataOrigin:
+    """Extract DataOrigin in a VO element
+
+    Parameters
+    ----------
+    vot_element : astropy.io.votable.tree.Info
+        VOTable Element to explore
+
+    Returns
+    -------
+    DataOrigin
+
+    Raises
+    ------
+    TypeError
+        input ``vot_element`` type is not supported
+    """
+    data_origin = DataOrigin()
+    if isinstance(vot_element, astropy.io.votable.tree.VOTableFile):
+        __extract_info_from_votable(vot_element, data_origin)
+    elif isinstance(vot_element, astropy.io.votable.tree.Resource):
+        __extract_info_from_resource(vot_element, data_origin)
+    elif isinstance(vot_element, astropy.io.votable.tree.Table):
```

Comment:
dang, while I was opening an issue for the previous comment, I remembered that `Table` is deprecated. Shouldn't this have issued a deprecationwarning? @pllim

Reply:
Hard to find out independently in a Python session because "tree" and "dataorigin" are not discoverable anymore. Pretty sure I used to be able to import "tree". Not sure what happened.

Reply:
My suspicion is that line is never executed but I cannot confirm.